File: fs-tests/storages.fs
module regiondeployer.tests.storage

open Xunit
open regiondeployer.cloudactions.storage
open Microsoft.Azure.Management.Fluent
open regiondeployer.cloudactions.resourcegroup
open System
open Microsoft.Azure.Management.Storage.Fluent
open regiondeployer.randomnamegenerator
open System.Linq
open regiondeployer.clouderror
open Xunit.Abstractions
open resourcegroup
open azureauthentication
open regiondeployer

type public when_a_storage_account_is_created(inputlogger:ITestOutputHelper) =    
    let maxtries = 10 //total edge case. there will never be 10 but there could be! /tableflip
    let grouptests = new when_it_creates_a_resource_group(inputlogger)
    let groupcontextsfield : string = grouptests.it_succeeds()     

    do
        configadapter.initialise();

    member val private authcontext = (new when_it_authenticates_to_azure(inputlogger)).it_succeeds()
    member private x.logger = inputlogger

    member public x.groupcontext = groupcontextsfield

    interface IDisposable with member this.Dispose() = 
        (grouptests :> IDisposable).Dispose()

    [<Fact>]
    ///returns a storage account in an isolated resourcegroup. if it was part of an existing group this would cause exceptions during test disposable.
    member public this.it_succeeds() : IStorageAccount =   
        try
            createrandomnamedstorageaccount(groupcontextsfield, this.logger.WriteLine)

        with
        | error1 -> 
            raiseclouderror error1  
            null

    [<Fact>]
    member public this.it_lives_in_the_input_resourcegroup() =
        let rec trymultipletimesgettingstorage (currenttries:int) : IStorageAccount =
            if currenttries < maxtries then
                let output = this.it_succeeds()

                if output.ResourceGroupName <> groupcontextsfield then //just in case I ran a test while there were existing farmers
                    (this.authcontext
                        .StorageAccounts
                        .DeleteById output.Id)

                    trymultipletimesgettingstorage (currenttries + 1)

                else
                    output

            else
                null

        try
            let objecttotest = trymultipletimesgettingstorage 0

            Assert.True(
                objecttotest.ResourceGroupName = groupcontextsfield
            )

        with
        | error1 -> 
            raiseclouderror error1  

File: tests/Storages.cs
using Xunit;
using Microsoft.Azure.Management.Fluent;
using Microsoft.Azure.Management.Storage.Fluent;
using System;
using Xunit.Abstractions;

namespace Tests
{
    public class WhenAStorageAccountIsCreated : IDisposable
    {
        private const int MaxTries = 10;
        private readonly WhenItCreatesAResourceGroup _groupTests;
        private readonly string _groupContextsField;
        private readonly ITestOutputHelper _logger;
        private readonly WhenItAuthenticatesToAzure _authContext;

        public WhenAStorageAccountIsCreated(ITestOutputHelper inputLogger)
        {
            _logger = inputLogger;
            _groupTests = new WhenItCreatesAResourceGroup(inputLogger);
            _groupContextsField = _groupTests.ItSucceeds();
            _authContext = new WhenItAuthenticatesToAzure(inputLogger).ItSucceeds();
            ConfigAdapter.Initialise();
        }

        public void Dispose()
        {
            _groupTests.Dispose();
        }

        [Fact]
        public IStorageAccount ItSucceeds()
        {
            try
            {
                return CreateRandomNamedStorageAccount(_groupContextsField, _logger.WriteLine);
            }
            catch (Exception error1)
            {
                RaiseCloudError(error1);
                return null;
            }
        }

        [Fact]
        public void ItLivesInTheInputResourceGroup()
        {
            IStorageAccount objectToTest = TryMultipleTimesGettingStorage(0);
            Assert.True(objectToTest.ResourceGroupName == _groupContextsField);
        }

        private IStorageAccount TryMultipleTimesGettingStorage(int currentTries)
        {
            if (currentTries < MaxTries)
            {
                var output = ItSucceeds();
                if (output.ResourceGroupName != _groupContextsField)
                {
                    _authContext.StorageAccounts.DeleteById(output.Id);
                    return TryMultipleTimesGettingStorage(currentTries + 1);
                }
                else
                {
                    return output;
                }
            }
            else
            {
                return null;
            }
        }

        private void RaiseCloudError(Exception error)
        {
            // Handle the error appropriately
            throw new InvalidOperationException("Cloud operation failed", error);
        }

        private IStorageAccount CreateRandomNamedStorageAccount(string resourceGroupName, Action<string> logAction)
        {
            // Implementation for creating a storage account with a random name
            logAction("Creating storage account");
            // Simulated creation logic
            return new StorageAccountSimulator(resourceGroupName);
        }
    }

    // Simulated classes for demonstration
    public class ConfigAdapter
    {
        public static void Initialise()
        {
            // Configuration initialization logic
        }
    }

    public class StorageAccountSimulator : IStorageAccount
    {
        public string ResourceGroupName { get; }
        public string Id { get; }

        public StorageAccountSimulator(string resourceGroupName)
        {
            ResourceGroupName = resourceGroupName;
            Id = Guid.NewGuid().ToString();
        }
    }
}
